### PR TITLE
Add "Attach to Github Issue" similar to Jira plugin

### PR DESCRIPTION
**Pull request by niklabh**
## Summary
Add "Attach to Github Issue" similar to Jira plugin
Here is the flow for implementations:
- Click Attach to Github Issue in context menu (shown when user connects his github account with plugin installed) on non system messages
Screenshot from 2019-09-11 13-26-14
- Attach to github modal will open
Screenshot from 2019-09-11 13-26-30
- You can search the issue. The query to github is is:open is:issue assignee:{connected_user} archived:false [org:{organisation}] [search term]
Screenshot from 2019-09-11 13-26-47
- Comment is attached to github
Screenshot from 2019-09-11 13-37-28

## Issue
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/119

### CODE REVIEW

**Comment by levb** [comment]
This is an awesome addition, @niklabh! Thank you for the contribution!

File: server/api.go [hunk 1]
```diff
@@ -480,6 +491,100 @@ func (p *Plugin) getYourPrs(w http.ResponseWriter, r *http.Request) {
 	w.Write(resp)
 }
 
+func (p *Plugin) searchIssues(w http.ResponseWriter, r *http.Request) {
+	config := p.getConfiguration()
```
Comment (mickmister) [inline server/api.go]: Nit: We're not checking for the correct HTTP method here. It's not being done for other endpoints in the plugin either, so it's probably not a big deal.

File: webapp/src/components/github_issue_selector/github_issue_selector.jsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+import debounce from 'debounce-promise';
+import AsyncSelect from 'react-select/async';
+
+import {getStyleForReactSelect} from 'utils/styles';
+import Client from 'client';
+
+const searchDebounceDelay = 400;
+
+export default class GithubIssueSelector extends Component {
```
Comment (mickmister) [inline webapp/src/components/github_issue_selector/github_issue_selector.jsx]:
```suggestion
export default class GithubIssueSelector extends PureComponent {
```

File: webapp/src/components/github_issue_selector/github_issue_selector.jsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        if (!this.props.required) {
+            return true;
+        }
+
+        const valid = this.props.value && this.props.value.toString().length !== 0;
+        this.setState({invalid: !valid});
+        return valid;
+    };
+
+    render = () => {
```
Comment (mickmister) [inline webapp/src/components/github_issue_selector/github_issue_selector.jsx]: Nit: I believe it's mostly convention, but prefer `render() {` syntax over `render = () => {` for the render method declaration.

File: webapp/src/components/github_issue_selector/github_issue_selector.jsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        const serverError = this.state.error;
+        let errComponent;
+        if (this.state.error) {
+            errComponent = (
+                <p className='alert alert-danger'>
+                    <i
+                        className='fa fa-warning'
+                        title='Warning Icon'
+                    />
+                    <span> {serverError.toString()}</span>
```
Comment (mickmister) [inline webapp/src/components/github_issue_selector/github_issue_selector.jsx]: Nit: unnecessary whitespace here
```suggestion
                    <span>{serverError.toString()}</span>
```

File: webapp/src/components/github_issue_selector/index.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import {connect} from 'react-redux';
+
+import GithubIssueSelector from './github_issue_selector';
+
+const mapStateToProps = () => ({});
+
+export default connect(mapStateToProps, null, null, {withRef: true})(GithubIssueSelector);
```
Comment (mickmister) [inline webapp/src/components/github_issue_selector/index.js]: Is this file necessary if we are not injecting any state or actions into the component?

File: webapp/src/components/modals/attach_comment_to_issue/index.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import {bindActionCreators} from 'redux';
+
+import {getPost} from 'mattermost-redux/selectors/entities/posts';
+
+import {closeAttachCommentToIssueModal, attachCommentToIssue} from 'actions';
+
+import AttachCommentToIssue from './attach_comment_to_issue';
+
+const mapStateToProps = (state) => {
+    const postId = state['plugins-github'].attachCommentToIssueModalForPostId;
```
Comment (mickmister) [inline webapp/src/components/modals/attach_comment_to_issue/index.js]: Can we import the plugin id, rather than hardcoding here?

File: webapp/src/components/modals/attach_comment_to_issue/index.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    const post = getPost(state, postId);
+
+    return {
+        visible: state['plugins-github'].attachCommentToIssueModalVisible,
```
Comment (mickmister) [inline webapp/src/components/modals/attach_comment_to_issue/index.js]: Import plugin id for this line as well

File: webapp/src/components/post_menu_actions/attach_comment_to_issue/index.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+import AttachCommentToIssuePostMenuAction from './attach_comment_to_issue';
+
+const mapStateToProps = (state, ownProps) => {
+    const post = getPost(state, ownProps.postId);
+    const systemMessage = post ? isSystemMessage(post) : true;
+
+    return {
+        isSystemMessage: systemMessage,
+        connected: state['plugins-github'].connected,
```
Comment (mickmister) [inline webapp/src/components/post_menu_actions/attach_comment_to_issue/index.js]: Import plugin id for this line

File: webapp/src/components/setting.jsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
```
Comment (mickmister) [inline webapp/src/components/setting.jsx]: Can we include the copyright header here? I see it's missing in a few other files as well. Curious how the linter didn't catch this.

File: webapp/src/components/setting.jsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        return (
+            <div className='form-group'>
+                <label
+                    className='control-label'
+                    htmlFor={inputId}
+                >
+                    {label}
+                </label>
+                {required &&
+                <span
```
Comment (mickmister) [inline webapp/src/components/setting.jsx]: Can we wrap the `span` in parens since it's multiline?

File: server/api.go (hunk 1 above)
Comment (niklabh) [reply server/api.go]: added method check

File: webapp/src/components/setting.jsx (hunk 9 above)
Comment (niklabh) [reply webapp/src/components/setting.jsx]: added everywhere

File: webapp/src/components/post_menu_actions/attach_comment_to_issue/index.js (hunk 8 above)
Comment (niklabh) [reply webapp/src/components/post_menu_actions/attach_comment_to_issue/index.js]: done

File: webapp/src/components/modals/attach_comment_to_issue/index.js (hunk 7 above)
Comment (niklabh) [reply webapp/src/components/modals/attach_comment_to_issue/index.js]: done

File: webapp/src/components/modals/attach_comment_to_issue/index.js (hunk 6 above)
Comment (niklabh) [reply webapp/src/components/modals/attach_comment_to_issue/index.js]: done

File: webapp/src/components/github_issue_selector/index.js (hunk 5 above)
Comment (niklabh) [reply webapp/src/components/github_issue_selector/index.js]: removed

File: webapp/src/components/github_issue_selector/github_issue_selector.jsx (hunk 4 above)
Comment (niklabh) [reply webapp/src/components/github_issue_selector/github_issue_selector.jsx]: done

File: webapp/src/components/github_issue_selector/github_issue_selector.jsx (hunk 3 above)
Comment (niklabh) [reply webapp/src/components/github_issue_selector/github_issue_selector.jsx]: done

File: webapp/src/components/setting.jsx (hunk 9 above)
Comment (mickmister) [reply webapp/src/components/setting.jsx]: Thanks for cleaning this up!

**Comment by hanzei** [comment]
@DHaussermann gentle reminder to review this

**Comment by DHaussermann** [comment]
Thanks for the contribution @niklabh. This addition to the functionality is great!
No issues found with the work in this PR. It is behaving as expected. 

@aaronrothschild I do see some opportunities to add to this further that would improve the experience. Let me know of any of this would be worth adding and I can create separate issues.   
1. When I attach a comment, I should get feedback that this was done successfully (Similar integrations post a reply with such a message and a link but, a simpler way to provide feedback wold work as well) 
2. In the search results, ideally we would return the issue number in addition to the title to cover cases where there titles are similar. Though the issue numbers could overlap across multiple repos within the same list. 
3. This design is limited to adding comments where the user is the assignee. There would also be value in adding comments to other issues as well. Would we need a more complex design to handle cases where you select an issue and don'y have permissions to comment?

**Comment by aaronrothschild** [comment]
@niklabh see my comments inline
> @aaronrothschild I do see some opportunities to add to this further that would improve the experience. Let me know of any of this would be worth adding and I can create separate issues.
> 
> 1. When I attach a comment, I should get feedback that this was done successfully (Similar integrations post a reply with such a message and a link but, a simpler way to provide feedback wold work as well)

Yes, I've had a few instances where the confirmation is re-assuring - certainly during the first experience.  I agree with @DHaussermann , the utility of having the link referenced within the conversation is pretty critical to the long-term user experience.  Over time, without the reference to the attached issue, a critical piece of development conversations can get lost.  it also serves the dual purpose of confirming that it was successful in attaching to the issue.  I'm 4/5 we should add this before merge.

> 2. In the search results, ideally we would return the issue number in addition to the title to cover cases where there titles are similar. Though the issue numbers could overlap across multiple repos within the same list.

This request has come up from a few different customers that use the equivalent function in the Jira plugin.  Displaying the issue numbers clears ambiquity and lets the user be confident they chose the right option/value.  Not sure how much effort is involved.

> 3. This design is limited to adding comments where the user is the assignee. There would also be value in adding comments to other issues as well. Would we need a more complex design to handle cases where you select an issue and don'y have permissions to comment?

Hmmm.  Not sure about this one, could we even accomodate a scenario where we couldn't post as the user (b/c permissions)?

**Comment by mickmister** [comment]
@aaronrothschild This PR is already merged. Adding the link in a post will have to be a separate PR.
